wfe/web-front-end.go

```diff
@@ -192,29 +192,77 @@ func marshalIndent(v interface{}) ([]byte, error) {
 	return json.MarshalIndent(v, "", "  ")
 }
 
+func (wfe *WebFrontEndImpl) relativeEndpoint(request *http.Request, endpoint string) string {
+	var result string
+	proto := "http://"
+	host := request.Host
+
+	// If the request was received via TLS, use `https://` for the protocol
+	if request.TLS != nil {
+		proto = "https://"
+	}
+
+	// Allow upstream proxies  to specify the forwarded protocol. Allow this value
+	// to override our own guess.
+	if specifiedProto := request.Header.Get("X-Forwarded-Proto"); specifiedProto != "" {
+		proto = fmt.Sprintf("%s://", specifiedProto)
+	}
+
+	// Default to "localhost" when no request.Host is provided. Otherwise requests
+	// with an empty `Host` produce results like `http:///acme/new-authz`
+	if request.Host == "" {
+		host = "localhost"
+	}
+
+	if wfe.BaseURL != "" {
+		result = fmt.Sprintf("%s%s", wfe.BaseURL, endpoint)
+	} else {
+		result = fmt.Sprintf("%s%s%s", proto, host, endpoint)
```

Comment:
Nit: This would be nicer as `net.URL{Scheme: ..., Host: ..., Path: ...}.String()`. I'm fine leaving the `BaseURL != ""` case as-is, since (a) that's how the code is written today, and (b) we'll eventually delete that branch.